### PR TITLE
Add Bash and Zsh command-line completion for flutter

### LINKED ISSUE #18988: flutter bash completion
It would be really nice to have a bash completion package so we don't always have to type the full commands in flutter

### PR DESCRIPTION
**Pull request by gspencergoog**
This uses @kevmoo's completion package to do command line completion for flutter, and a new command "bash-completion" (with alias "zsh-completion") that will output the necessary shell script setup code, and adds the hidden command "completion" that does the actual completion.

Because it adds a dependency, I also had to do `flutter update-packages --force-upgrade`.

Fixes #18988.

### CODE REVIEW

**Comment by kevmoo** [comment]
I'm not a good person to review this. Thoughts, @Hixie ?

**Comment by gspencergoog** [comment]
OK, I really just wanted to make sure I was using your package the way it was intended.

**Comment by kevmoo** [comment]
It looks okay to me. Were you able to wire up completion and use it?

**Comment by gspencergoog** [comment]
Yep, it works.  The performance is "reasonable".  I wouldn't call it fast, but it's definitely usable.

File: packages/flutter_tools/lib/src/commands/shell_completion.dart [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+  @override
+  final String description =
+    'Output command line shell completion setup scripts.\n\n'
+    'This command prints the flutter command line completion script for Bash and Zsh. To use it,\n'
+    'redirect the output into a file and follow the instructions in the output to install it in\n'
+    'your shell environment. Once added, your shell will be able to complete flutter commands and\n'
+    'options.\n\n'
+    'The Flutter tool anonymously reports feature usage statistics and basic crash reports to help improve\n'
+    'Flutter tools over time. See Google\'s privacy policy: https://www.google.com/intl/en/policies/privacy/';
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/shell_completion.dart]: This description seems out of place here?

File: packages/flutter_tools/lib/src/runner/flutter_command_runner.dart [hunk 2]
```diff
@@ -177,6 +178,25 @@ class FlutterCommandRunner extends CommandRunner<Null> {
...
+      Command<Null> command = commands[error.commands.first];
+      for (String commandName in error.commands.skip(1)) {
+        command = command.subcommands[commandName];
+      }
+
+      command.usageException(error.message);
+      return null;
+    }
+  }
+
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/runner/flutter_command_runner.dart]: can you explain the purpose of this change?

**Comment by Hixie** [comment]
tests?

**Comment by gspencergoog** [comment]
Added the missing test file (forgot to `git add`).

Also, added support for specifying an output file instead of only writing to stdout.

File: packages/flutter_tools/lib/src/runner/flutter_command_runner.dart (hunk 2 above)
Comment (Hixie) [reply packages/flutter_tools/lib/src/runner/flutter_command_runner.dart]: Can you include a big comment that explains all this? Thanks.

File: packages/flutter_tools/lib/src/commands/shell_completion.dart [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      'install it in your shell environment. Once it is sourced, your shell will be able to\n'
+      'complete flutter commands and options.';
+
+  @override
+  final List<String> aliases = <String>['zsh-completion'];
+
+  @override
+  bool get shouldUpdateCache => false;
+
+  /// Return null to disable tracking of the `bash-completion` command.
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/shell_completion.dart]: s/tracking/analytics recording/

File: packages/flutter_tools/lib/src/commands/shell_completion.dart [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+      stdout.write(generateCompletionScript(<String>['flutter']));
+      return;
+    }
+
+    final File outputFile = fs.file(argResults.rest.first);
+    if (outputFile.existsSync() && !argResults['overwrite']) {
+      throwToolExit(
+          'Output file ${outputFile.path} already exists, will not overwrite. '
+          'Use --overwrite to force overwriting existing output file.',
+          exitCode: 1);
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/shell_completion.dart]: trivial nit: trailing comma and newline before )

**Comment by Hixie** [comment]
LGTM

cc @tvolkert who may also have opinions

File: packages/flutter_tools/lib/src/commands/shell_completion.dart (hunk 3 above)
Comment (gspencergoog) [reply packages/flutter_tools/lib/src/commands/shell_completion.dart]: Done, and fixed in config.dart too (where a similar comment was).

File: packages/flutter_tools/lib/src/commands/shell_completion.dart (hunk 4 above)
Comment (gspencergoog) [reply packages/flutter_tools/lib/src/commands/shell_completion.dart]: Done.

File: packages/flutter_tools/lib/src/runner/flutter_command_runner.dart (hunk 2 above)
Comment (Hixie) [reply packages/flutter_tools/lib/src/runner/flutter_command_runner.dart]: I think I was reviewing an older copy of the PR.

**Comment by Solido** [comment]
Is there any process to activate it ? 
My Zsh cannot complete Flutter cmd.
Thanks

**Comment by gspencergoog** [comment]
Yes, in the help for the `bash-completion` command, it gives instructions.  Basically, you run it as so:
```bash
flutter bash-completion flutter_completion.sh
```
and it will write a completion setup script. You then `source flutter_completion.sh` in your shell startup (.zshrc for Zsh).

You might also place the `flutter_completion.sh` in a `completion.d` directory renamed as `flutter` if you have other completion scripts there already that are sourced automatically.

**Pull request merged**